### PR TITLE
feat: Commands Interface [only for inter-platform communication calls]

### PR DESCRIPTION
**Pull request by utmishra**
- Based on the discussion with Marc, Leo and Luca, we have decided not to overcomplicate the Commands architecture and keep only inter-platform communication calls under Commands.
- **Structure**:
```
export const CommunicationInterface {
  dispatch(command: <CommandType>, payload?: [Optional]): CommandResponse // contains on(eventName, callback)}
}
```

The `CommandResponse` returns methods which can be subscribed to adding callback logic (Similar to [Shopify's action calls](https://shopify.dev/docs/apps/tools/app-bridge/actions/toast))

### CODE REVIEW

File: packages/app-bridge/src/commands/AssetViewer.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
+
+import { Command, CommandResponse } from '../types';
+
+export interface AssetViewer {
+    dispatch(command: Command): CommandResponse[Command.AssetViewer];
```
Comment (triggertoo) [inline packages/app-bridge/src/commands/AssetViewer.ts:6]:
```suggestion
    dispatch(command: Command.AssetViewer): CommandResponse[Command.AssetViewer];
```

File: packages/app-bridge/src/types/Command.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
+
+import { Asset } from './Asset';
+
+export enum Command {
+    AssetViewer = 'AssetViewer',
+    AssetChooser = 'assetChooser',
+}
```
Comment (triggertoo) [inline packages/app-bridge/src/types/Command.ts:8]: it's probably better to avoid enums in these case, since it's dangerous to export this through types as it's not only a type (compile team) but also javascript code (runtime).

File: packages/app-bridge/src/AppBridgeBlock.ts [hunk 3]
```diff
@@ -29,6 +29,10 @@ export interface AppBridgeBlock extends AppBridgeBase {
 
     addAssetIdsToBlockAssetKey(key: string, assetIds: number[]): Promise<void>;
 
+    /**
+     * @deprecated Use dispatch(`openAssetViewer`) command instead
+     * This will be removed in version 4.0.0 of @frontify/app-bridge
+     */
```
Comment (SamuelAlev) [inline packages/app-bridge/src/AppBridgeBlock.ts:35]: Keep it while making the implementation run `dispatch("openAssetViewer")` so no breaking changes

File: packages/app-bridge/src/commands/AssetChooser.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
+
+import { Command, CommandResponse } from '../types';
+
+export interface AssetChooser {
```
Comment (SamuelAlev) [inline packages/app-bridge/src/commands/AssetChooser.ts:5]: type not interface

File: packages/app-bridge/src/commands/AssetViewer.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
+
+import { Command, CommandResponse } from '../types';
+
+export interface AssetViewer {
```
Comment (SamuelAlev) [inline packages/app-bridge/src/commands/AssetViewer.ts:5]: type not interface

File: packages/app-bridge/src/commands/AssetViewer.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
+
+import { Command, CommandResponse } from '../types';
```
Comment (SamuelAlev) [inline packages/app-bridge/src/commands/AssetViewer.ts:3]:
```suggestion
import type { Command, CommandResponse } from '../types';
```

File: packages/app-bridge/src/tests/AppBridgeBlockStub.ts [hunk 7]
```diff
@@ -4,14 +4,14 @@ import mitt, { Emitter } from 'mitt';
...
+import { EmitterEvents } from '../types/Emitter';
+import { PrivacySettings } from '../types/PrivacySettings';
 import { AssetDummy } from './AssetDummy';
-import { UserDummy } from './UserDummy';
-import { ColorPaletteDummy } from './ColorPaletteDummy';
-import { ColorDummy } from './ColorDummy';
 import { BulkDownloadDummy } from './BulkDownloadDummy';
-import { PrivacySettings } from '../types/PrivacySettings';
+import { ColorDummy } from './ColorDummy';
+import { ColorPaletteDummy } from './ColorPaletteDummy';
```
Comment (SamuelAlev) [inline packages/app-bridge/src/tests/AppBridgeBlockStub.ts:13]: Reduce noise, please set up your ide to prevent that or revert the changes

File: packages/app-bridge/src/tests/AppBridgeThemeStub.ts [hunk 8]
```diff
@@ -3,10 +3,6 @@
 import mitt, { Emitter } from 'mitt';
 import { SinonStubbedInstance, spy, stub } from 'sinon';
 
-import { EmitterEvents } from '../types/Emitter';
-import { AppBridgeTheme } from '../AppBridgeTheme';
-import { mergeDeep } from '../utilities';
```
Comment (SamuelAlev) [inline packages/app-bridge/src/tests/AppBridgeThemeStub.ts:8]: Reduce noise, please set up your ide to prevent that or revert the changes

File: packages/app-bridge/src/types/Command.ts (hunk 2 above)
Comment (utmishra) [reply packages/app-bridge/src/types/Command.ts]: Removed this and used direct values

File: packages/app-bridge/src/tests/AppBridgeThemeStub.ts (hunk 8 above)
Comment (utmishra) [reply packages/app-bridge/src/tests/AppBridgeThemeStub.ts]: Reverted

File: packages/app-bridge/src/commands/AssetViewer.ts (hunk 6 above)
Comment (utmishra) [reply packages/app-bridge/src/commands/AssetViewer.ts]: Using as type now

File: packages/app-bridge/src/types/Command.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+export type CommandResponse = {
+    AssetViewer: {
+        on(event: CommandEvents.AssetViewerOpened, callback: () => void): void;
+        on(event: CommandEvents.AssetViewerClosed, callback: () => void): void;
+    };
+    AssetChooser: {
+        on(event: CommandEvents.AssetChooserOpened, callback: (selectedAssets: Asset[]) => void): void;
+        on(event: CommandEvents.AssetChooserClosed, callback: () => void): void;
+    };
+};
```
Comment (triggertoo) [inline packages/app-bridge/src/types/Command.ts:21]:
```suggestion
export type CommandResponse = {
    AssetViewer: {
        on(event: 'assetViewerOpened', callback: () => void): void;
        on(event: 'assetViewerClosed', callback: () => void): void;
    };
    AssetChooser: {
        on(event: 'assetChooserOpened', callback: (selectedAssets: Asset[]) => void): void;
        on(event: 'assetChooserClosed', callback: () => void): void;
    };
};
```

**Review by triggertoo — COMMENTED** [review]
no enum needed

**Review by triggertoo — APPROVED** [review]
lgtm 👍

**Comment by utmishra** [comment]
**Update**
- The type definitions for the AssetChooser/AssetViewer commands needed update so they would be compatible with/similar to the existing AppBridge methods
- **Pending/Open question**: 
    - Hooks are publicly implemented in the AppBridge. How to leverage Command _interface_ to add the action wrapper inside these hooks

**Comment by triggertoo** [comment]
not sure about that [last commit](https://github.com/Frontify/brand-sdk/pull/442/commits/95d88fe491a25ab0036aa687791a72fd511c2f6b), what's the idea of having an interface for `AssetViewer` instead using function overload on the `invoke` function within the app bridge?

**Comment by triggertoo** [comment]
> **Update**
> 
> * The type definitions for the AssetChooser/AssetViewer commands needed update so they would be compatible with/similar to the existing AppBridge methods
> * **Pending/Open question**:
>   
>   * Hooks are publicly implemented in the AppBridge. How to leverage Command _interface_ to add the action wrapper inside these hooks
> 
> cc: @SamuelAlev @triggertoo

From my understanding, we don't change the way how hooks are being used. We just update the hooks to use the new command pattern.

**Comment by utmishra** [comment]
Closed in favor of #460
